modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java

```diff
@@ -683,6 +699,70 @@ private JSONArray _getItemsActionsJSONArray(
 			});
 	}
 
+	private String _getNestedFields(
+			String apiUrl, Set<ObjectEntry> fdsFieldsSet)
+		throws Exception {
+
+		if (fdsFieldsSet == null) {
+			return apiUrl;
+		}
+
+		String nestedFields = StringPool.BLANK;
+
+		int nestedFieldsDepth = 1;
+
+		JSONArray fdsFieldsJSONArray = JSONUtil.toJSONArray(
+			fdsFieldsSet,
+			(ObjectEntry objectEntry) -> {
+				Map<String, Object> properties = objectEntry.getProperties();
+
+				return JSONUtil.put(
+					"fieldName", String.valueOf(properties.get("name")));
+			});
+
+		for (int i = 0; i < fdsFieldsJSONArray.length(); i++) {
+			JSONObject fdsFieldJSONObject = fdsFieldsJSONArray.getJSONObject(i);
+
+			String[] fieldNameArray = StringUtil.split(
+				fdsFieldJSONObject.getString("fieldName"), CharPool.PERIOD);
+
+			if (fieldNameArray.length > 1) {
+				String[] fieldsName = new String[fieldNameArray.length - 1];
+
+				System.arraycopy(
+					fieldNameArray, 0, fieldsName, 0,
+					fieldNameArray.length - 1);
+
+				for (String fieldName : fieldsName) {
+					nestedFields = StringUtil.add(nestedFields, fieldName);
+				}
+
+				if (fieldNameArray.length > nestedFieldsDepth) {
+					nestedFieldsDepth = fieldNameArray.length - 1;
+				}
+			}
+		}
+
+		if (nestedFields.equals(StringPool.BLANK)) {
+			return apiUrl;
+		}
+
+		StringBundler sb = new StringBundler(5);
+
+		sb.append(apiUrl);
+		sb.append("?nestedFields=");
+		sb.append(
+			StringUtil.replaceLast(
+				nestedFields, CharPool.COMMA, StringPool.BLANK));
```

Comment:
To simplify code here, you can create an empty `ArrayList` on start, add in loop, and `StringUtil.merge()` on end.

Reply:
I think that the order is not important. Just want to prevent sending multiple times the same root property.

x.e. with a selection like `catalog.*`,  `catalog.name` and `catalog.id` we could end up adding up to 3 `catalog` entries for the `nestedFields` parameter.

Reply:
Anyway, I've found a way of iterating over the fieldsSet `Set` :crossed_fingers: